### PR TITLE
Add a SearchControl component and reuse accross the UI

### PR DESCRIPTION
**Pull request by youknowriad**
Builds on top of #32922 

The need for a reusable search control UI component became clear over time. It's now being used in three places: Inserter, Block Manager and Template Part placeholder state.

I tried to make its API close to all the other `*Control` components. I think we can still push it further by reusing `BaseControl` and providing `help`, `hideLabelFromVision` props support and things like that.

**Testing instructions**

 - Check that the search in the three places above work properly.

### CODE REVIEW

**Review by jasmussen — APPROVED** [review]
Looks great to me:

yo

There's a small bottom border present in the template part search picker which doesn't seem necessary anymore, this one:

Screenshot 2021-06-23 at 20 09 38

We can probably remove that.

File: packages/components/src/search-control/index.js [hunk 1]
```diff
@@ -8,44 +8,43 @@ import classnames from 'classnames';
...
+ */
+import { VisuallyHidden, Button } from '../';
+
+function SearchControl( {
 	className,
 	onChange,
 	value,
 	label,
-	placeholder,
+	placeholder = __( 'Search' ),
```
Comment (sarayourfriend) [inline packages/components/src/search-control/index.js:25]: I think we were trying to avoid introducing new usages of `wordpress/i18n` into `wordpress/components` if possible (preferrably we want to remove existing usages if possible 🤔)

What do you think about making `placeholder` a required prop?

**Review by sarayourfriend — COMMENTED** [review]
I think converting this to use BaseControl is a good idea.

Not sure about `wordpress/i18n` it may be an inextricable dependency of `components` 🤷 

What do you think about adding some tests for this component?

File: packages/components/src/search-control/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: to be honest, that doesn't sound great to me, most components would need some default wording in them and these need to be translatable, what's the reasoning behind the need to remove this dependency?

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: If someone wanted to adopt `@wordpress/components` without adopting a whole new i18n solution (for example an already existing app that uses `react-i18n` but wants to use `@wordpress/components`).

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: How do third-party UI component libraries deal with this?

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: most third-party UI component libraries don't include the "composite" or higher-level components that wp/components includes and therefore don't run into this problem. Even Material UI which has a ton of higher-level components just requires you to pass labels and placeholders manually.

Material UI _does_ handle l10n however: https://material-ui.com/guides/localization/

But their `TextField` doesn't include a default label, for example. It's just required: https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/TextField/TextField.js#L90

Chakra doesn't include any translatable strings at all as far as I can tell.

However because of the higher-level components like ColorPicker in wp/components I don't think we actually have that option (aside from leveraging tree-shaking to avoid importing wp/i18n if you just avoid using those higher-level components for example).

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: So what's the best plan for us in your opinion?

 - Remove i18n where possible and accept that there are some places where it's unavoidable?
 - Use it without restrictions because it's unavoidable?
 - Figure out a way to inject the i18n helpers?

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: > Remove i18n where possible and accept that there are some places where it's unavoidable?

That's probably the simplest option. With tree-shaking it achieves the best result for most external use cases that don't involve the higher-level components.

Alternatively:

> Figure out a way to inject the i18n helpers?

That's an interesting idea but would require adding a new provider for it, which I'm not against necessarily but adding a required provider isn't necessarily ideal (we've been able to avoid it thus far).

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: Ok makes sense to me though I'm not sure what to do in this component, we have two strings: "Search" and "Reset search". One can be removed by making the placeholder mandatory but not the other.

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: We can't add a `resetLabel` prop?

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: > We can't add a resetLabel prop?
 
I don't like that strategy personally because we'll be introducing props (especially mandatory props that we repeat all over the places) just to solve a technical issue.

I agree that using i18n makes our components less reusable (forcing you to rely on `@wordpress/i18n` for your app) but it sounds to me that it may be still our best option until we figure out a better strategy that don't force us to add weird props like that (props might be a solution but I feel we should think about this problem more holistically first)

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: Interesting.

Isn't the worst case scenario that if the consumer is repeating things all over the place they just create their own wrapper around `SearchInput` that adds the labels they want (using the i18n solution of their choice even)?

```jsx
import { SearchInput } from '@wordpress/components';
import { translate } from 'my-favorite-i18n-solution';

const MySearchInput = ( { label = translate( 'Search' ), resetLabel = translate( 'Reset search' ), ...props } ) => (
  <SearchInput label={ label } resetLabel={ resetLabel } { ...props } />
);
```

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: > Isn't the worst case scenario that if the consumer is repeating things all over the place they just create their own wrapper around SearchInput that adds the labels they want (using the i18n solution of their choice even)?

Potentially if this override is adding value, in this case, it seems it's just a solution to get around a technical problem preventing you from adding these sane defaults to the component itself. Basically what you're suggestion is two "components" packages: one to be consumed by third-parties and one to be consumed by WordPress. Given our main focus is WordPress and related third-party usage (plugins, themes...), it sounds a bit overkill for us to do so.

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: > Basically what you're suggestion is two "components" packages

I've definitely proposed this before 🙂 I still don't think it's a bad idea to separate the primitives from the higher-level components but I understand that it's not a popular opinion 😀 

> Given our main focus is WordPress and related third-party usage

That's fair. I think there are probably some significant use cases (like someone wanting to use Gutenberg outside of a WP context) where being able to extract the i18n solution entirely into something abstract that can be replaced ad-hoc via a provider isn't an unrealistic goal (this is more achievable now that we have `useI18n` from `react-i18n`, though it is yet to be used widely within the project).

However, I understand that we have to focus our resources wisely and that non-WP usages of Gutenberg are probably on the back burner.

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: I think non-WP usage is good to keep in my mind and I always support making it easier. But in this particular situation, I think for now advocating for `i18n` usage in addition to components is not very bad to warrant adding more complexity to the other aspect of our work (WP usage). I wouldn't be against exploring clever solutions to inject optional i18n providers in the future.

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: This issue also doesn't seem specific to the components package, the block-editor package is another good example here, and one that would require a lot of "messages" and replacing all of these with props sound very unrealistic to me.

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: Sounds good to me.

How about we update this component to use `react-i18n` at least?

And thanks for this discussion, it's always a pleasure to discuss these issues with you and revisiting them often helps realize more and more levels of details of potential solutions.

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: > How about we update this component to use react-i18n at least?

I'm not yet familiar with this package yet :). Quickly checking it, it seems the idea is to replace the i18n global functions with hooks right? I assume this has some advantages, would you mind clarifying them?

The downside here would be that we won't be able to use it in any place other than components, I'm pretty sure we have a number of these in our code base that would need to be rethought.

Comment (sarayourfriend) [reply packages/components/src/search-control/index.js:25]: One advantage is that you could theoretically wrap your usage of the WordPress components with the `I18nProvider` from that package and pass your own i18n solution as `__` and the other functions.

> The downside here would be that we won't be able to use it in any place other than components

Correct `react-i18n` is explicitly for React components.

@sirreal and @swissspidy may be able to elucidate the benfits better than myself given they wrote the original impementation of react-i18n. Whatever those benefits are we should add them to the README for that package.

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: To just share my opinion here, I wouldn't mind switching slowly to  `react-i18n` but It seems we need to audit non react i18n usage to check whether such a big refactor will pay off in the end or not.

Comment (sirreal) [reply packages/components/src/search-control/index.js:25]: Issue https://github.com/WordPress/gutenberg/issues/19210 does a thorough job of describing the shortcomings of `@wordpress/i18n` that `@wordpress/react-i18n` was designed to address. In summary:

- `@wordpress/i18n` relies on a single store that shares state. This is problematic in cases like server-side rendering (SSR) where we may respond to requests in multiple languages.
- `@wordpress/i18n` is not "reactive," i.e. it cannot notify component of changes to trigger updates. If we want to change the language of a page using `@wordpress/i18n` with a language switcher it requires a full reload or some external mechanism to force React to re-render in order to update translations.

In summary, `@wordpress/react-i18n` is a thin wrapper that makes the `@wordpress/i18n` API better fit the React paradigm.

Comment (youknowriad) [reply packages/components/src/search-control/index.js:25]: I've been thinking more, we do use translations a lot in block strings (titles, descriptions...) and all of these are out of the React rendering, it seems we'll never be able to use `react-i18n` fully.

Comment (sirreal) [reply packages/components/src/search-control/index.js:25]: This is likely worth creating a dedicated issue, some exploration has already happened in #27973.

Blocks will indeed have that limitation at this time for certain properties, although I don't believe it applies to `@wordpress/components` which are implemented on top of React.

**Comment by youknowriad** [comment]
Add the BaseControl usage to the component, which means now, it supports help and hideLabelFromVision (which I made true by default for this control, it seemed like a better default for a Search control)

**Comment by youknowriad** [comment]
I could use some help here with the react native failures @mchowning @geriux

**Comment by geriux** [comment]
> I could use some help here with the react native failures @mchowning @geriux

Sure! I'll check it out 👍

**Comment by geriux** [comment]
@youknowriad 👋 it is currently failing because of the missing `style.native.scss` with the styles of the native variant of `packages/components/src/search-control/index.native.js`

Adding that file with the styles fixes the issue.

**Comment by youknowriad** [comment]
@geriux Thanks for the hint, I think I fixed it (moved the styles from the inserter native style to the search control one), let's see what the tests say.
